lib/srv/db/proxy_test.go

```diff
@@ -0,0 +1,192 @@
+	// sure the connection succeeds.
+	psql, err := testCtx.postgresClientAddr(ctx, proxy.Address(), "alice", "postgres", "postgres", "postgres")
+	require.NoError(t, err)
+	require.NoError(t, psql.Close(ctx))
+}
+
+// TestProxyProtocolMySQL ensures that clients can successfully connect to a
+// MySQL database when Teleport is running behind a proxy that sends a proxy
+// line.
+func TestProxyProtocolMySQL(t *testing.T) {
+	ctx := context.Background()
+	testCtx := setupTestContext(ctx, t, withSelfHostedMySQL("mysql"))
+	t.Cleanup(func() { testCtx.Close() })
+	go testCtx.startHandlingConnections()
+
+	testCtx.createUserAndRole(ctx, t, "alice", "admin", []string{"root"}, []string{types.Wildcard})
+
+	// Point our proxy to the Teleport's MySQL listener.
+	proxy, err := newTestProxy(testCtx.mysqlListener.Addr().String())
+	require.NoError(t, err)
+	t.Cleanup(func() { proxy.Close() })
+	go proxy.Serve()
+
+	// Connect to the proxy instead of directly to MySQL listener and make
+	// sure the connection succeeds.
+	mysql, err := testCtx.mysqlClientAddr(proxy.Address(), "alice", "mysql", "root")
+	require.NoError(t, err)
+	require.NoError(t, mysql.Close())
+}
+
+// testProxy is tcp passthrough proxy that sends a proxy-line when connecting
+// to the target server.
+type testProxy struct {
+	listener net.Listener
+	target   string
+	closeCh  chan (struct{})
+	log      logrus.FieldLogger
+}
+
+// newTestProxy creates a new test proxy that sends a proxy-line when
+// proxying connections to the provided target address.
+func newTestProxy(target string) (*testProxy, error) {
+	listener, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &testProxy{
+		listener: listener,
+		target:   target,
+		closeCh:  make(chan struct{}),
+		log:      logrus.WithField(trace.Component, "test:proxy"),
+	}, nil
+}
+
+// Address returns the proxy listen address.
+func (p *testProxy) Address() string {
+	return p.listener.Addr().String()
+}
+
+// Serve starts accepting client connections and proxying them to the target.
+func (p *testProxy) Serve() error {
+	for {
+		clientConn, err := p.listener.Accept()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		go func() {
+			if err := p.handleConnection(clientConn); err != nil {
+				p.log.WithError(err).Error("Failed to handle connection.")
+			}
+		}()
+	}
+}
+
+// handleConnection dials the target address, sends a proxy line to it and
+// then starts proxying all traffic b/w client and target.
+func (p *testProxy) handleConnection(clientConn net.Conn) error {
+	serverConn, err := net.Dial("tcp", p.target)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer serverConn.Close()
+	errCh := make(chan error, 2)
+	go func() { // Client -> server.
+		defer clientConn.Close()
+		defer serverConn.Close()
+		// Write proxy-line first and then start proxying from client.
+		err := p.sendProxyLine(clientConn, serverConn)
```

Comment:
Should this be in the Server -> Client goroutine instead?
Otherwise, you have a race between sending proxyline and the actual server sending data

Reply:
Hmm, I don't think so? We first send proxy line to the server and then start sending whatever the client has to send?

Reply:
You're right, forgot proxy line is a server-facing thing